app/src/main/java/com/android/sample/mlUtils/MlCoach.kt

```diff
@@ -0,0 +1,204 @@
+   * @return A `CoachFeedback` object containing feedback for the exercise.
+   */
+  fun getFeedBackSingleExercise(
+      data: List<List<MyPoseLandmark>>,
+      exerciseCriterion: ExerciseCriterion,
+      preambleCriterions: ExerciseCriterion,
+      isTimeBased: Boolean = false,
+      numberOfRepetition: Int = 0,
+  ): CoachFeedback {
+    val data_preambleActived =
+        data.filter { sample -> ExerciseFeedBack.assessLandMarks(sample, preambleCriterions).first }
+
+    var exerciseDuration = 0L
+    if (data_preambleActived.isNotEmpty()) {
+      val firstTimeStamp = data_preambleActived.first().first().timeStamp
+      val lastTimeStamp = data_preambleActived.last().first().timeStamp
+      exerciseDuration = (lastTimeStamp - firstTimeStamp) / 1000L
+    }
+    // get the list of comments
+    val assessedExercise =
+        data_preambleActived.map { sample ->
+          ExerciseFeedBack.assessLandMarks(sample, exerciseCriterion = exerciseCriterion)
+        }
+    val nbComments = assessedExercise.fold(0, { acc, (boolean, list) -> acc + list.size })
+    val allComments =
+        assessedExercise
+            .flatMap { (sampleIsSucceed, commentList) -> commentList }
+            .groupingBy { it }
+            .eachCount()
+            .toList()
+            .map { (comment, count) -> Pair(comment, count / nbComments.toFloat()) }
+            .sortedBy { (_, freq) -> freq }
+
+    var successRate = 0F
+    val jointFeedbackSet: MutableSet<JointFeedback> = mutableSetOf()
+    val listOfAdvice =
+        allComments.forEach { (comment, freq) ->
+          if (comment == AngleCriterionComments.SUCCESS) {
+            successRate = freq
+          } else {
+            jointFeedbackSet.add(JointFeedback(comment = comment.description, rate = freq))
+          }
+        }
+
+    return if (isTimeBased) {
+      CoachFeedback(
+          commentSet = jointFeedbackSet,
+          successRate = successRate,
+          repOrDuration = exerciseDuration.toInt(),
+          "s",
+          exerciseCriterion = exerciseCriterion)
+    } else {
+      CoachFeedback(
+          commentSet = jointFeedbackSet,
+          successRate = successRate,
+          repOrDuration = numberOfRepetition,
+          "repetitions",
+          exerciseCriterion = exerciseCriterion)
+    }
+  }
+}
+
+/**
+ * Counts the number of alternates in the assessed poses.
+ *
+ * @param assessedPoses A list of lists of Booleans representing the assessed poses.
+ * @return The number of alternates in the assessed poses.
+ */
+fun countAlternates(assessedPoses: List<List<Boolean>>): Int {
+  // Each column of the matrix represent the detection state of a given exercise criterion eg: for
+  // the pushups, the 1st column is the detection of the up position and the 2nd column is the
+  // detection of the down position
+  // lists have the same length, remove when there is more than one true for the same index
```

Comment:
Again just for my curiosity, what do you mean when you say remove when there is more than one true for the same index?

Reply:
Indeed this is not really clear. 
Assume you have two lists A and B where A is the list where the low pushup posiiton is detected and B the list when the up pushup position is detected. 
When I mean detected, it implies preabmle = True for a given moment in the list A or B.
As you cannot be at the same time in position A (low) and B(up) when doing a pushup, you remove these from the computation.
Hope this is clear